qemu/tests/live_snapshot_chain.py

```diff
@@ -78,94 +79,101 @@ def do_operate(params, key_word):
-    if vm.is_alive():
-        vm.destroy()
-    if params.get("remove_snapshot_images"):
-        for index, image in enumerate(snapshot_chain):
+        error.context("Do base files check", logging.info)
+        snapshot_chain_backward = snapshot_chain[:]
+        snapshot_chain_backward.reverse()
+
+        for index, image in enumerate(snapshot_chain_backward):
             image_params = params.object_params(image)
-            if index != 0:
-                image = qemu_storage.QemuImg(
-                    image_params, data_dir.get_data_dir(), image)
-                image.remove()
+            if image_params.get("check_base_image"):
+                vm.destroy()
+                vm.create(params=image_params)
+                vm.verify_alive()
+
+                session = vm.wait_for_login(timeout=timeout)
+                if image_params.get("file_create"):
+                    for file in md5_value[image]:
+                        md5 = session.cmd_output(md5_cmd % file)
+                        if md5 != md5_value[image][file]:
+                            error_message = "File %s in image %s changed " %\
+                                            (file, image)
+                            error_message += "from '%s' to '%s'(md5)" %\
+                                             (md5_value[image][file], md5)
+                            raise error.TestFail(error_message)
+                    files_check = session.cmd(file_check_cmd % file_dir)
+                    if files_check != files_in_guest[image]:
+                        error_message = "Files in image %s is not as expect:" %\
+                                        image
+                        error_message += "Before shut down: %s" %\
+                            files_in_guest[image]
+                        error_message += "Now: %s" % files_check
+                        raise error.TestFail(error_message)
+                if image_params.get("image_check"):
+                    image = qemu_storage.QemuImg(
+                        image_params, data_dir.get_data_dir(), image)
+                    image.check_image(image_params, data_dir.get_data_dir())
+                session.close()
+    finally:
+        error.context("Remove snapshot images", logging.info)
+        if vm.is_alive():
+            vm.destroy()
+        if params.get("remove_snapshot_images"):
+            missing = []
+            for index, image in enumerate(snapshot_chain):
+                image_params = params.object_params(image)
+                if index != 0:
+                    image = qemu_storage.QemuImg(
+                        image_params, data_dir.get_data_dir(), image)
+                    if not os.path.exists(image.image_filename):
+                        missing.append(image.image_filename)
+                    image.remove()
+            test.assertFalse(missing, "The %s snapshot(s) are not present on"
+                             " filesystem." % missing)
```

Comment:
If the test has been already failed before, this would hide the truth.

Reply:
You are right I should move the missing assertion to the test body and leave the `finally` block only to cleanup. I'll add it in v2.

Anyway the finally block is necessary as this test leaves (many) files behind in case of failure.